landoapi/api/treestatus.py

```diff
@@ -0,0 +1,609 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import collections
+import copy
+import datetime
+import functools
+import json
+import logging
+
+from typing import (
+    Any,
+    Callable,
+    Optional,
+)
+
+import pytz
+import sqlalchemy as sa
+
+from connexion import ProblemException
+from flask import g
+
+from landoapi import auth
+from landoapi.cache import cache
+from landoapi.models.treestatus import (
+    DEFAULT_TREE,
+    Log,
+    StatusChange,
+    StatusChangeTree,
+    Tree,
+    load_last_state,
+)
+from landoapi.storage import db
+
+logger = logging.getLogger(__name__)
+
+
+TREE_SUMMARY_LOG_LIMIT = 5
+
+CombinedTree = collections.namedtuple(
+    "CombinedTree",
+    ["tree", "message_of_the_day", "tags", "status", "reason", "log_id", "model"],
+)
+
+
+def get_combined_tree(
+    tree: Tree,
+    tags: Optional[list[str]] = None,
+    status: Optional[str] = None,
+    reason: Optional[str] = None,
+    log_id: Optional[int] = None,
+) -> CombinedTree:
+    """Combined view of the Tree.
+
+    This also shows status, reason and tags from last Tree Log.
+    """
+    result = copy.deepcopy(DEFAULT_TREE)
+    result.update(tree.to_dict())
+
+    if tags is not None:
+        result["tags"] = json.loads(tags)
+
+    if status is not None:
+        result["status"] = status
+
+    if reason is not None:
+        result["reason"] = reason
+
+    if log_id is not None:
+        result["log_id"] = log_id
+
+    result["model"] = tree
+
+    return CombinedTree(**result)
+
+
+def combinedtree_as_dict(tree: CombinedTree) -> dict[str, Any]:
+    """Convert a `CombinedTree` to a dict.
+
+    Removes the `model` field as part of the conversion.
+    """
+    return {field: value for field, value in tree._asdict().items() if field != "model"}
+
+
+def result_object_wrap(f: Callable) -> Callable:
+    """Wrap the value returned from `f` in a result object.
```

Comment:
Is this meant to be a "result dictionary"?

Reply:
The output from the API will be JSON, which usually uses `object`. I'll switch it to `dict`.